wpep0008.md

```diff
@@ -0,0 +1,50 @@
+| Key | Value |
+| ---: | :--- |
+| **Title** | Wazo integration tests framework |
+| **Number** | 0008 |
+| **Date** | 2023-08-28 |
+| **Authors** | Jesse Sopel <jesse.sopel@wazo.io>, Sebastien Duthil <sduthil@wazo.io> |
+| **Status** | draft |
+
+## Abstract
+
+Wazo integration tests: Explicit the direction of evolution of integration tests
+
+## Motivation
+
+The integration test framework was implemented in Wazo before the usage of
+pytest. Pytest offers a very Pythonic way to organize tests and fixtures, and we
+want to use this system more and more, since it offers more flexibility,
+readability, performance and reusability.
+
+## State of the art in Wazo Platform
+
+Integration tests are organised around `AssetLaunchingTestCase`, inheriting from
+the `unittest` testing framework. This organization forces tests to be organized
+around a test class, which may not be as flexible as we want.
+
+## Rejected propositions
+
+* None, this proposition emerged from a refactor proposition that looked promising.
+
+## Proposition
+
+* Migrate all Wazo Platform projects to Pytest (looking at you wazo-confd!)
+* Use Pytest fixtures system as much as possible for organizing tests and fixtures.
+
+The `AssetLaunchingTestCase` class [was
+refactored](https://github.com/wazo-platform/wazo-test-helpers/pull/89) to be
+usable in more Pytest friendly patterns, while keeping retrocompatibility with
+the former style of tests.
```

Comment:
Could be more specific on what such a migration would involve, such as including an example of what a good pytest-oriented test case would look like and how a typical unittest-based test case could be converted to a pytest-based one.

Reply:
(I guess simplifying the agid PR to an essential self-contained example that demonstrates some aspects of the new AssetLaunchingTestCase version)

Reply:
The only major changes were to split out the logic into another class so we didn't have to create a TestCase that extends from the test helper. The decorator was just a helper to make it easier to work with the unusual class level behavior of this class.

If this is the direction we are going. It might be nice to plan to eventually migrate away from said class level usage of attributes in `AssetLaunchingTestCase`? It might be a good idea to look into as it would allow us to simply use functools.cached_property instead of the Descriptor I had to write.

Reply:
Also, it should be "while keeping backwards-compatible with"